docs/sources/lids/_DRAFT-Index-Gateway-Sharding.md

```diff
@@ -0,0 +1,99 @@
+
+The index gateway service can be run in "simple mode", where an index gateway instance is responsible for handling, storing and returning requests for all indices for all tenants, or in "ring mode", where an instance is responsible for a subset of tenants instead of all tenants.
+
+On top of that, in order to achieve redundancy as well as spreading load, the index gateway ring uses a replication factor of 3.
+
+This means, before an index gateway client makes a request to the index gateway server, it first hashes the tenant ID and then requests a replication set for that hash from the index gateway ring. Due to the fixed replication factor (RF), the replication set contains three server addresses. On every request, a random server from that list is picked to then execute the request on.
+
+## Problem Statement
+
+The current strategy of sharding by tenant ID and having a replication factor of 3 fails in the long run, because even when running lots of index gateways, only a maximum of three could be utilized by a single tenant. 
+
+Another problem is that the RF is fixed and the same for all tenants, independent of their actual size in terms of log volume or query rate.
+
+## Goals
+
+The goal of this document is to find a better sharding mechanism for the index gateway, so that there are no boundaries for scaling the service horizontally.
+
+* The sharding needs to account for the "size" of a tenant.
+* A single tenant needs to be able to utilize more than three index gateways.
+
+## Proposals
+
+### Proposal 0: Do nothing
+
+If we do not improve the sharding mechanism for the index gateways and leave it as it is, it will become more and more difficult to serve metadata queries for large tenant in a reasonable amount of time, proportionally to the demand for these queries.
+
+### Proposal 1: Dynamic replication factor
+
+Instead of using a fixed replication factor of 3, the RF can be derived from the amount of active members in the index gateway ring. That means that the RF would be a percentage of the available gateway instances. For example, a ring with 12 instances and 30% replication utilization would result in a RF of 3 (`floor(12*0.3)`). Scaling up to 18 instances would result in a RF of 5.
+
+This approach would solve the problem of horizontal scaling. However, it does not solve the problem of different tenant sizes. It also fails to ensure replication for availability when running a small number of instances, unless there is a fixed lower value for the RF. It also tends to over-shard data in large index gateway deployments.
+
+### Proposal 2: Fixed per-tenant replication factor
+
+Adding a random shard ID (e.g. `shard-0`, `shard-1`, ... `shard-n`) to the tenant ID allows to utilize a certain amount of `n` instances. The amount of shards can be implemented as a per-tenant override. However, this approach results in non-deterministic hash keys.
+
+### Proposal 3: Shard by index files
```

Comment:
By usage, do you mean the size of the index?

Reply:
Ok, not size of the index, but how often the index is queried

Reply:
Size plays a part, but I'm referring more to the underlying _work_ that needs to be done. For instance, tenant A may have 100TB/day of ingest and does a lot of querying. Even with sharding, we'd "hotspot" access to this daily index for said tenant.

Reply:
I like this proposal but have a couple of suggestions to avoid hot-spotting and make it less agnostic of underlying storage:
* Use table number/bucket instead of file names to avoid jumping around index gateways since each file change also changes the file names. This also keeps the index client agnostic of how the index is managed on the storage.
* Consider hashing with tenant id as well with the table number to avoid hot-spotting the same index gateways for all the tenants querying data in the same time interval.

This would help better distribute the load.

Reply:
this is how we calculate table number/buckets. You can ignore table name since it won't add any value here
https://github.com/grafana/loki/blob/2208ea7147d4ac99071cd9d468990d97e12bc447/pkg/storage/stores/tsdb/manager.go#L284